<a href="https://colab.research.google.com/github/pachterlab/varseek-examples/blob/main/vk_count.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [vk count](https://github.com/pachterlab/varseek) demonstration
Perform variant screening on scRNA-seq data with vk count, using a [10x PBMC 1k dataset](https://www.10xgenomics.com/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0) against the sample mutation database from the vk ref tutorial. 

Written by Joseph Rich.
___

### Install varseek, and import all packages

In [1]:
try:
    import varseek as vk
except ImportError:
    print("varseek not found, installing...")
    !pip install -U -q varseek

In [2]:
import os
import subprocess
import shutil
import anndata as ad
import numpy as np

import varseek as vk

### Define important paths

In [3]:
# vk count out directory
vk_count_out_dir = os.path.join("data", "varseek_count_out")
reference_dir = os.path.join("data", "reference")

# vk ref out directory and files - downloaded if not already present
vk_ref_dir = os.path.join("data", "varseek_ref_out_sample")
vcrs_index = os.path.join(vk_ref_dir, "vcrs_index.idx")
vcrs_t2g = os.path.join(vk_ref_dir, "vcrs_t2g.txt")

# reference genome index for pseudobam validation (only used when pseudobam_validation=True). Our variants are HGVSc (transcript-level), annotated against the GRCh37 Ensembl release 93 cDNA (see vk_ref.ipynb), so the pseudobam reference is that cDNA indexed with kallisto.
reference_genome_cdna = os.path.join(reference_dir, "ensembl_grch37_release93", "Homo_sapiens.GRCh37.cdna.all.fa")
reference_genome_index = os.path.join(reference_dir, "ensembl_grch37_release93", "cdna_index.idx")

# fastq directories - fastqs_dir downloaded if not already present, and fastqs_processed_dir created with fastp if not already present
fastqs_dir = os.path.join("data", "pbmc_1k_v3_fastqs")
do_fastp = False
technology = "10xv3"

# general
pseudobam_validation = False
threads = 8

In [ ]:
if pseudobam_validation:
    vk_count_out_dir += "_pseudobam_correction"

### Download the vk ref index and t2g files if they do not already exist. See vk_ref.ipynb for more details.

In [4]:
if not os.path.exists(vcrs_index):
    vcrs_index_url = "https://caltech.box.com/shared/static/6cnqes4ewieq464m3nugsctq5k0miub1.idx"
    vk.utils.download_box_url(vcrs_index_url, output_folder=vk_ref_dir, output_file_name="vcrs_index.idx")

if not os.path.exists(vcrs_t2g):
    vcrs_t2g_url = "https://caltech.box.com/shared/static/z23alf7g73vegnqv6b0nptzm7uolu1dk.txt"
    vk.utils.download_box_url(vcrs_t2g_url, output_folder=vk_ref_dir, output_file_name="vcrs_t2g.txt")

### (Only if `pseudobam_validation=True`): Provide the reference cDNA for pseudobam validation
Pseudobam validation re-aligns each read with `kallisto quant` against a standard reference and drops reads whose true alignment coordinate is inconsistent with the variant its VCRS represents. `vk count` builds this reference index automatically (with `kb ref --workflow custom`) from `sequences` when `reference_genome_index` does not yet exist. Because our variants are HGVSc (transcript-level), `sequences` is the GRCh37 Ensembl 93 cDNA, so here we only need to make sure that cDNA is present.

In [5]:
# vk count builds the pseudobam reference index from the cDNA (`sequences`) when needed, so here we just make
# sure the GRCh37 Ensembl 93 cDNA is present (same file used as `sequences` in vk_ref.ipynb).
if pseudobam_validation and not os.path.exists(reference_genome_cdna):
    cdna_dir = os.path.dirname(reference_genome_cdna)
    !gget ref -w cdna -r 93 --out_dir {cdna_dir} -d human_grch37
    !gunzip {reference_genome_cdna}.gz

### Download the PBMC fastq dataset

In [6]:
if not os.path.exists(fastqs_dir) or len(os.listdir(fastqs_dir)) == 0:
    !mkdir -p data && \
        cd data && \
        curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar && \
        tar -xvf pbmc_1k_v3_fastqs.tar && \
        rm pbmc_1k_v3_fastqs.tar

### (Recommended): Process the fastq data - as an example, we will use [fastp](https://github.com/OpenGene/fastp)
Note: fastp was not designed for single-cell data, and as such it requires careful thought to ensure correct processing. The procedure belows works on 10xv3 fastq files, but may need to be modified for other datasets. The function vk.utils.perform_fastp_trimming_and_filtering (called internally by vk.count) handles these nuances. While vk count would handle this step, we call it outside of vk count to demonstrate the process for clarity.

In [7]:
if do_fastp and shutil.which("fastp"):
    fastqs_processed_dir = os.path.join(fastqs_dir, "filtered")
    if not os.path.exists(fastqs_processed_dir) or len(os.listdir(fastqs_processed_dir)) == 0:
        vk.fastqpp(
            fastqs = fastqs_dir,
            quality_control_fastqs = True,
            technology = technology,
            threads = threads,
            cut_front = True,
            cut_tail = True,
            disable_length_filtering = True,
            dont_eval_duplication = True,
            disable_trim_poly_g = True,
            quality_control_fastqs_out_dir = fastqs_processed_dir
        )
        
        fastqs_dir = fastqs_processed_dir
else:
    print("fastp is not installed. Skipping fastq pre-processing")

fastp is not installed. Skipping fastq pre-processing


### Run varseek count

This will run the following commands:
- `varseek fastqpp`: Preprocess the fastq files. By default, does nothing.
- `kb count` (variant reference): Perform variant screening on fastq data utilizing kb count's pseudoalignment algorithm. Variant data is stored in an Anndata object as as cell/sample x variant matrix.
- `varseek clean`: Process the output of kb count. By default, this will threshold variant counts and ensure that there is agreement for each read between the gene of the variant to which the read aligned during the variant reference pseudoalignment and the gene to which the read aligned during the "normal" reference genome pseudoalignment.
- `varseek summarize`: Produces a text file summarizing some high-level insights from the variant screening process.

In [16]:
vk_count_output_dict = vk.count(
    fastqs=fastqs_dir,
    index=vcrs_index,
    t2g=vcrs_t2g,
    technology=technology,
    out=vk_count_out_dir,
    threads=threads,
    # quality_control_fastqs=True, cut_front=True, cut_tail=True, disable_length_filtering=True, dont_eval_duplication=True, disable_trim_poly_g=True,  # equivalent to the fastp step above
    pseudobam_validation=pseudobam_validation, reference_genome_index=reference_genome_index, sequences=reference_genome_cdna,
)

15:55:11 - INFO - Removing index files from fastq files list, as they are not utilized in kb count with technology 10XV3
15:55:11 - INFO - Setting length_required to 59 if fastqpp is run
15:55:11 - INFO - Skipping vk fastqpp because there was no use for it
15:55:11 - INFO - Running kb count with command: kb count -t 32 -k 59 -i data/varseek_ref_out_sample/vcrs_index.idx -g data/varseek_ref_out_sample/vcrs_t2g.txt -x 10XV3 --h5ad -o data/varseek_count_out_pseudobam_correction/kb_count_out_vcrs --overwrite --union --num --mm data/pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz data/pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz data/pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz data/pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz
[2026-07-17 15:55:18,828]    INFO [count] Using index data/varseek_ref_out_sample/vcrs_index.idx to generate BUS file to data/varseek_count_out_pseudobam_correction/kb_count_out_vcrs from
[2026-07-17 15:55:18,828]    INFO [count]         d

In [17]:
print(f"Find the processed adata object for further analysis in {vk_count_output_dict['adata_path']}")

Find the processed adata object for further analysis in /home/jrich/Desktop/varseek-examples/data/varseek_count_out_pseudobam_correction/adata_cleaned.h5ad


## Load in the anndata object

In [11]:
adata = ad.read_h5ad(vk_count_output_dict["adata_path"])
adata

AnnData object with n_obs × n_vars = 1214 × 119
    var: 'vcrs_id', 'seq_ID', 'mutation', 'nucleotide_positions', 'actual_variant', 'start_variant_position', 'end_variant_position', 'variant_source', 'variant_count', 'variant_detected', 'number_obs'

## Keep only cells with non-zero counts

In [12]:
adata = adata[:, np.array((adata.X != 0).sum(axis=0)).ravel() > 0]
adata

View of AnnData object with n_obs × n_vars = 1214 × 69
    var: 'vcrs_id', 'seq_ID', 'mutation', 'nucleotide_positions', 'actual_variant', 'start_variant_position', 'end_variant_position', 'variant_source', 'variant_count', 'variant_detected', 'number_obs'

In [13]:
adata.var.head()

,vcrs_id,seq_ID,mutation,nucleotide_positions,actual_variant,start_variant_position,end_variant_position,variant_source,variant_count,variant_detected,number_obs
variant,,,,,,,,,,,
ENST00000368716:c.413T>A,ENST00000368716:c.413T>A,ENST00000368716,c.413T>A,413,T>A,413,413,transcriptome,4.0,True,2
ENST00000368716:c.436G>A,ENST00000368716:c.436G>A,ENST00000368716,c.436G>A,436,G>A,436,436,transcriptome,2.0,True,1
ENST00000368716:c.368G>T,ENST00000368716:c.368G>T,ENST00000368716,c.368G>T,368,G>T,368,368,transcriptome,6.0,True,3
ENST00000368716:c.391C>A,ENST00000368716:c.391C>A,ENST00000368716,c.391C>A,391,C>A,391,391,transcriptome,2.0,True,1
ENST00000368716:c.383C>A,ENST00000368716:c.383C>A,ENST00000368716,c.383C>A,383,C>A,383,383,transcriptome,8.0,True,4


## pseudobam analysis

In [4]:
adata_raw = ad.read_h5ad(f"{vk_count_out_dir}/adata_cleaned.h5ad")
adata_pseudobam = ad.read_h5ad(f"{vk_count_out_dir}_pseudobam_correction/adata_cleaned.h5ad")

In [5]:
import numpy as np
from scipy import sparse

# Align cells and genes
all_obs = adata_raw.obs_names.union(adata_pseudobam.obs_names)
all_var = adata_raw.var_names.union(adata_pseudobam.var_names)

# Reindex both AnnData objects
adata_raw = adata_raw[all_obs, all_var]
adata_pseudobam = adata_pseudobam[all_obs, all_var]

# Binarize
X1 = (adata_raw.X > 0).astype(np.uint8)
X2 = (adata_pseudobam.X > 0).astype(np.uint8)

# Convert to CSR if sparse
if not sparse.issparse(X1):
    X1 = sparse.csr_matrix(X1)
if not sparse.issparse(X2):
    X2 = sparse.csr_matrix(X2)

# Drop columns that are empty in BOTH matrices
keep_cols = np.asarray((X1.getnnz(axis=0) + X2.getnnz(axis=0)) > 0).ravel()
X1 = X1[:, keep_cols]
X2 = X2[:, keep_cols]
gene_names = adata_raw.var_names[keep_cols]

# Compare
diff = X1 != X2

# Number of differing entries
n_entry_diff = diff.nnz
union_nnz = (X1.maximum(X2)).nnz

# Rows (cells) with at least one difference
rows_with_diff = np.asarray(diff.getnnz(axis=1) > 0).ravel()
n_row_diff = rows_with_diff.sum()

# Columns (genes/VCRSs) with at least one difference
cols_with_diff = np.asarray(diff.getnnz(axis=0) > 0).ravel()
n_col_diff = cols_with_diff.sum()

print(f"Differing entries: {n_entry_diff:,} / {union_nnz:,} ({100*n_entry_diff/union_nnz:.2f}%)")
print(f"Rows with differences: {n_row_diff:,} / {X1.shape[0]:,} ({n_row_diff / X1.shape[0]:.2%})")
print(f"Columns with differences: {n_col_diff:,} / {X1.shape[1]:,} ({n_col_diff / X1.shape[1]:.2%})")

Differing entries: 174 / 404 (43.07%)
Rows with differences: 154 / 1,214 (12.69%)
Columns with differences: 17 / 69 (24.64%)


In [16]:
# Categorize every read the pseudobam validation kept vs. dropped (requires pseudobam_validation = True).
import re, subprocess, collections
import pandas as pd

kb_dir = os.path.join(f"{vk_count_out_dir}_pseudobam_correction", "kb_count_out_vcrs")
bam = os.path.join(kb_dir, "pseudobam_qc", "quant_group_0", "pseudoalignments.bam")
bus_df_path = os.path.join(kb_dir, "bus_df.parquet")

if not (os.path.exists(bam) and os.path.exists(bus_df_path)):
    print("Run vk count with pseudobam_validation = True first - this analysis reads the pseudobam intermediates.")
else:
    bus_df = pd.read_parquet(bus_df_path)

    # ENST -> gene (ENSG) and gene_biotype, parsed from the cDNA fasta headers
    enst2gene, gene_biotype = {}, {}
    with open(reference_genome_cdna) as fh:
        for line in fh:
            if not line.startswith(">"):
                continue
            enst = line[1:].split()[0].split(".")[0]
            g = re.search(r"gene:(\S+)", line); gb = re.search(r"gene_biotype:(\S+)", line)
            if g:
                enst2gene[enst] = g.group(1).split(".")[0]
                if gb:
                    gene_biotype[enst2gene[enst]] = gb.group(1)

    # only reads with a matchable FASTQ header could be checked against the pseudobam
    checkable = bus_df[bus_df["fastq_header"].notna()].copy()
    target_file = os.path.join(kb_dir, "pseudobam_qc", "_target_headers.txt")
    checkable["fastq_header"].drop_duplicates().to_csv(target_file, index=False, header=False)

    # stream the BAM, keeping only the target reads' (read, transcript) pairs (low memory)
    awk = r"""FNR==NR{t[$1]=1;next} ($1 in t){split($3,a,"."); print $1"\t"a[1]}"""
    cmd = f"samtools view -@ {threads} -F 4 {bam} | awk -F'\t' '{awk}' {target_file} - | sort -u"
    out = subprocess.run(cmd, shell=True, capture_output=True, text=True).stdout
    aln = pd.DataFrame([l.split("\t") for l in out.splitlines() if l], columns=["fastq_header", "enst"])
    aln_by_read = aln.groupby("fastq_header")["enst"].apply(set).to_dict()

    def vcrs_transcripts(names):
        items = list(names) if hasattr(names, "__iter__") and not isinstance(names, str) else [names]
        return [pp.split(":")[0].split(".")[0] for v in items for pp in str(v).split(";") if ":" in pp]

    cat = collections.Counter()
    per_tx = collections.defaultdict(lambda: [0, 0, 0])  # transcript -> [kept, dropped_same_gene, dropped_different_gene]
    for _, r in checkable.iterrows():
        vts = vcrs_transcripts(r["gene_names"]); vgenes = {enst2gene.get(t) for t in vts}
        tx = vts[0] if vts else "?"
        aligned = aln_by_read.get(r["fastq_header"])
        if aligned is None:
            cat["kept: read did not align to the transcriptome"] += 1
            per_tx[tx][0] += 1
        elif any(t in aligned for t in vts):
            cat["kept: read aligned to its variant's transcript"] += 1
            per_tx[tx][0] += 1
        else:
            same = bool(vgenes & {enst2gene.get(t) for t in aligned})
            cat["DROPPED: same gene, different isoform/pseudogene" if same else "DROPPED: different gene entirely"] += 1
            per_tx[tx][1 if same else 2] += 1

    n_unchecked = int(bus_df["fastq_header"].isna().sum())
    print(f"pseudobam-checkable reads: {len(checkable)}  (another {n_unchecked} reads had no FASTQ-header match and were kept unchecked)\n")
    for k in ["kept: read aligned to its variant's transcript",
              "kept: read did not align to the transcriptome",
              "DROPPED: same gene, different isoform/pseudogene",
              "DROPPED: different gene entirely"]:
        print(f"  {cat[k]:5d}  {k}")

    # gene symbols are not in the GRCh37 release-93 cDNA headers, so look them up (needs internet; falls back to "?")
    top = sorted(per_tx.items(), key=lambda kv: -sum(kv[1][1:]))[:8]
    gene2symbol = {}
    try:
        import gget
        ensgs = sorted({enst2gene.get(tx) for tx, _ in top} - {None})
        if ensgs:
            info = gget.info(ensgs, verbose=False)
            for g, row in info.iterrows():
                for col in ("primary_gene_name", "ensembl_gene_name"):  # per-gene fallback: either may be missing
                    if isinstance(row.get(col), str) and row[col]:
                        gene2symbol[g] = row[col]
                        break
    except Exception as e:
        print(f"\n(gene symbol lookup unavailable: {e})")

    print("\nTop variant-transcripts by dropped reads  (reads: original -> final, dropped = same-gene + different-gene):")
    for tx, (kept, same, diff) in top:
        dropped = same + diff
        original = kept + dropped
        g = enst2gene.get(tx)
        symbol = gene2symbol.get(g, "?")
        print(f"  {tx}  original={original:5d} final={kept:5d} dropped={dropped:5d} (same={same:4d} diff={diff:3d})  "
              f"gene={symbol} / {g} ({gene_biotype.get(g, '?')})")


pseudobam-checkable reads: 2826  (another 2828 reads had no FASTQ-header match and were kept unchecked)

   1664  kept: read aligned to its variant's transcript
    100  kept: read did not align to the transcriptome
    947  DROPPED: same gene, different isoform/pseudogene
    115  DROPPED: different gene entirely

Top variant-transcripts by dropped reads  (reads: original -> final, dropped = same-gene + different-gene):
  ENST00000424325  original=  866 final=   23 dropped=  843 (same= 799 diff= 44)  gene=RPL10 / ENSG00000147403 (protein_coding)
  ENST00000512805  original=   94 final=    0 dropped=   94 (same=  94 diff=  0)  gene=RACK1 / ENSG00000204628 (protein_coding)
  ENST00000417839  original=   95 final=   49 dropped=   46 (same=   0 diff= 46)  gene=? / ENSG00000203827 (protein_coding)
  ENST00000429416  original=   41 final=    0 dropped=   41 (same=  41 diff=  0)  gene=SMARCA4 / ENSG00000127616 (protein_coding)
  ENST00000309863  original=   21 final=    0 dropped=   21 (same

### What this shows

- **Most dropped reads are same-gene, not random.** Of the ~2,800 checkable reads, ~1,660 aligned to their variant's own transcript (kept). Of the ~1,060 dropped, **~89% aligned to a different isoform or pseudogene of the *same* gene** and only ~11% to a genuinely different gene — the removed reads are "a little off" (wrong isoform/paralog), not wildly misassigned.
- **One artifact dominates.** `ENST00000424325` = **RPL10**, a very highly expressed ribosomal protein, accounts for ~840 of the drops. Those reads pseudoalign across RPL10, its paralog RPL10L, and several RPL10 **pseudogenes** — the textbook single-cell false-positive pattern (paralog / pseudogene cross-mapping). The next cluster, `ENST00000512805` = RACK1/GNB2L1, is the same story. A "variant" with hundreds of counts on a housekeeping ribosomal gene in a healthy donor is almost certainly artifact, so removing these improves specificity.
- **Which matrix is more correct?** There's no orthogonal ground truth on this donor, but the validated matrix is the more *specific* one: the bulk of what it removes is paralog/pseudogene cross-mapping, which inflates the unvalidated matrix.
- **Two caveats.** (1) ~2,800 reads whose 10x barcode was error-corrected can't be matched back to a FASTQ header and are **kept unchecked** — pseudobam actually scrutinized only ~half the reads. (2) The default check is transcript-level only. Setting `check_alignment_position=True` additionally requires the read to land within ~50 bp of the variant; on this data that dropped a further ~736 reads (473 → 390 counts; 61 → 50 non-zero variants), i.e. many reads sit on the correct transcript but not at the variant locus — a stricter, more principled test.